src/sas/qtgui/Perspectives/Fitting/FittingUtilities.py

```diff
@@ -576,21 +576,14 @@ def plotPolydispersities(model, disp_models):
         return plots
     # test for model being a sasmodels.sasview_model.SasviewModel?
     for name, pars in model.dispersion.items():
-        if 0 == pars['width']:
-            continue # no actual polydispersity
-        disp_args = pars.copy()
-        disp_type = disp_args.pop('type', None)
-        if disp_type is None:
-            continue # show an error? "poly type ... not found"
-        disp_func = disp_models[disp_type](**disp_args)
-        center = model.params[name]
-        # same as in sasmodels.weights.Dispersion
-        lb = center - disp_args['nsigmas']*disp_args['width']*center
-        ub = center + disp_args['nsigmas']*disp_args['width']*center
-        arr = disp_func.get_weights(center, lb, ub, relative = True)
+        if 0 == pars['width']: # no actual polydispersity
+            continue
+        _, xarr, yarr = model.get_weights(name)
+        if xarr is None or yarr is None: # param name not found
+            continue
```

Comment:
Unfortunately, there is a case (the Maier-Saupe distribution) where a "width" of zero indicates that the particle is unoriented, and so the it ought to return a distribution.  While we do not treat this correctly yet (various code takes shortcuts when width is zero), we may eventually, so don't make this assumption here. 

Also, the get_weights function will never return None for the array, only sequences of length one if there is no distribution.  

And again, you are not referencing `pars` in the loop, so no need to loop over it.

So something like:
```
    for name model.dispersion:
        xarr, yarr = model.get_weights(name)
        if len(xarr) > 1:
            data1d = Data1D(x=xarr, y=yarr)
            ...
```

Reply:
Thanks for the feedback!
Regarding *model.get_weights()* there is a problem now: It can't detect if the polydispersity was turned off again as the previously calculated weights remain stored in the model. Even from the GUI point of view it is not possible to detect if a polydispersity has to be plotted - assuming that a width == 0 is a valid value.
For example, the ellipsoid model offers 4 parameters which can be polydisperse in theory. At the moment, all 4 are plotted. How to tell which one is actually 'active' polydisperse?
Can you give a hint, what should be changed (GUI or backend) in order to make this work?

Reply:
I don't see how weights remain stored in the model. `model.get_weights()` is calling  `SasviewModel._get_weights()` which calls `weights.get_weights()` each time and does not update `self`.  `get_weights()` always returns a distribution, though it may be only a single point of weight 1.  This is the monodisperse case that doesn't need to be plotted.  So the test for `len(xarr) <= 1` that you have should be enough.  

Maybe you need to clear the existing plots from the GUI before entering your loop?
 
Note that there are some potential threading issues that can cause problems.  There is only one `SasviewModel` object and it is used by the GUI, the calculator, and the fit engine, each of which is running in a separate thread.  The fit engine in particular is stuffing values into the model independent of the GUI, so the state within the model may not match the state displayed on the GUI.   Don't think that is happening here though.